README.md
# Screentasia

<p align="center">
  <img src="https://user-images.githubusercontent.com/74715917/284740018-918bcf55-87ed-4325-b384-f757f0349d34.jpg" width="350px">
</p>

# Level Up Your UI, Enjoyable & Simple

## #1 Best Screens UI Package

  

# ![](https://user-images.githubusercontent.com/74715917/284740021-7f34dfc1-72ef-4639-8079-fc8612254afb.gif) ✨ Give a star for the Repo ✨

#   

# Screentasia

  

A Flutter package that helps you create beautifully adaptive designs for all screens with a wealth of features.

  

## Features

  

*   🚀 Adaptive From: To adaptive sizes from different screens like Mobile, Tablet, Desktop
*   ⚡️ Adaptive Percentage: To customize Adaptive percentage for different screens sizes in percentage for the hole app or for a specific widget
*   ➗️ Percentage Sizes: Now you can customize width and height as percentages from screen
*   ❤️ As you want: You can disable Adaptive for specific screens just passing 0
*   🏋‍♂️️ No Media Query: No need for Media Query no more

  

## Usage

  

### Add dependency

  

Please check the latest version before installation.

If there is any problem with the new version, please use the previous version

  

```yaml
dependencies:
  flutter:
    sdk: flutter
  # add screentasia
  screentasia: ^{latest version}
```

  

### Add the following imports to your Dart code

  

```dart
import 'package:screentasia/screentasia.dart';
```

  

### Property

  

| Property | Type | Default Value | Description |
| ---| ---| ---| --- |
| adaptiveFrom | double | adaptiveFrom . mobile | Used to adapt sizes from chosen screen ex: if 500 width on a desktop is equal to 50% of the screen width, will be equal to 50% to all screens,also you can use custom screen with customScreen method |
| adaptivePercentage | AdaptivePercentage | AdaptivePercentage(mobile: 100, tablet: 100, desktop: 100) | For adaptive percentage in different screens, also can disable adaptive for specific screens with 0 value as parameter |
| builder | Function | null | Return widget that uses the library in a property (ex: MaterialApp's theme) |
| child | Widget | null | A part of the builder that its dependencies/properties don't use the library |
| rebuildFactor | Function | _default_ | Returns whether to rebuild or not when screen metrics change. |

  

**Note : You must either provide builder, child or both.**

  

### Initialize

  

### To Initialize the Tha Adaptive with you app

  

#### The first way (You must use it once in your app)

  

```scala
void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
   
    return ScreentasiaInit(
      adaptiveFrom: AdaptiveFrom.mobile, //optional for custom screen ex: customScreen(400,700)
      adaptivePercentage: AdaptivePercentage(mobile:100,tablet:100,desktop:100),//optional
      builder: (context , child) {
        return MaterialApp(
          debugShowCheckedModeBanner: false,
          title: 'Example',
          home: child,
        );
      },
      child: const HomePage(title: 'Example'),
    );
  }
}
```

  

### Examples

### In this example we will use only width for all situations to use adaptive values,Of course we can apply that concept in (Height,Font,Padding,etc)

  

### Example using Values

  

We can create adaptive value for (width,height,font,etc) based on value in specific device for example (Mobile,Tablet,Desktop) or specific device dimension ( width,height )

to apply adaptive value for all screen

ex: 250.w '250 width in mobile screen'

  

```haskell
class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return ScreentasiaInit(
      adaptiveFrom: AdaptiveFrom.mobile,
      adaptivePercentage:
          const AdaptivePercentage(mobile: 100, desktop: 100, tablet: 100),
      builder: (context, child) {
        return MaterialApp(
            debugShowCheckedModeBanner: false,
            title: 'Flutter screentasia',
            home: Scaffold(
                appBar: AppBar(
                  title: const Text('Flutter screentasia Demo  Page'),
                ),
                body: Container(
                  color: Colors.blue,
                  width: 250.w,
                )));
      },
    );
  }
}
```

#### Desktop

![](https://user-images.githubusercontent.com/74715917/284739887-daf52730-bed4-4783-ad2c-327163534b58.png)

#### Tablet

![](https://user-images.githubusercontent.com/74715917/284739894-556301bc-9db2-437b-a339-4a2aa4f9617e.png)

### Mobile

![](https://user-images.githubusercontent.com/74715917/284739900-9f3d85b1-6326-4211-96ef-bada89aa7654.png)

###   

### Example using Percentage

  

We can create adaptive value for (width,height, font,etc) based on percentage

for example we will use width percentage 'wb' in diffrent screen sizes

ex: 50wp '50 width percent in screen'

  

```plain
class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return ScreentasiaInit(
      adaptiveFrom: AdaptiveFrom.mobile,
      adaptivePercentage:
          const AdaptivePercentage(mobile: 100, desktop: 100, tablet: 100),
      builder: (context, child) {
        return MaterialApp(
            debugShowCheckedModeBanner: false,
            title: 'Flutter screentasia',
            home: Scaffold(
                appBar: AppBar(
                  title: const Text('Flutter screentasia Demo  Page'),
                ),
                body: Container(
                  color: Colors.blue,
                  width: 50.wp,
                )));
      },
    );
  }
}
```

#### Desktop

![](https://user-images.githubusercontent.com/74715917/284739968-6e8c7b61-10ab-44b7-941e-f41ecabce332.png)

#### Tablet

![](https://user-images.githubusercontent.com/74715917/284739973-eb2c9aa5-62cf-4e05-8f98-f1d51a49a175.png)

### Mobile

![](https://user-images.githubusercontent.com/74715917/284739976-d1e426a2-bf20-4a06-b5ce-ad565c6e55d7.png)

###   

### Example using Percentage in Specific Screen

  

We can create adaptive value for (width, height, font, etc) based on percentage

for example, we will use width percentage 'wb' in different screen sizes with different value percentages in specific screens

ex: 100wp '100 width per cent from Desktop Screen'

50wp '50 width per cent from Tablet Screen'

25wp '25 width per cent from Mobile Screen'

  

```haskell
class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return ScreentasiaInit(
      adaptiveFrom: AdaptiveFrom.mobile,
      adaptivePercentage:
          const AdaptivePercentage(mobile: 100, desktop: 100, tablet: 100),
      builder: (context, child) {
        return MaterialApp(
            debugShowCheckedModeBanner: false,
            title: 'Flutter screentasia',
            home: Scaffold(
                appBar: AppBar(
                  title: const Text('Flutter screentasia Demo  Page'),
                ),
                body: Container(
                    color: Colors.blue,
                    width: 100.wp.ap(
                        adaptivePercentage: const AdaptivePercentage(
                            desktop: 100, tablet: 50, mobile: 25)))));
      },
    );
  }
}
```

  

#### Desktop

![](https://user-images.githubusercontent.com/74715917/284739988-a889cf6d-5fe0-45a8-962f-a567f630f029.png)

#### Tablet

![](https://user-images.githubusercontent.com/74715917/284739991-a7b962c5-c468-4a3d-86ac-f09ffda110f4.png)

### Mobile

![](https://user-images.githubusercontent.com/74715917/284739994-90bfc9ba-90d0-490d-ae62-d786c703bad9.png)

###   

###   

### API

  

#### Pass the dp size of the design draft

  

```scss
    Screentasia().setWidth(540)  (dart sdk>=2.6 ) another way (540.w) //Adapted to screen width
    Screentasia().setHeight(200) (dart sdk>=2.6 ) another way (200.h) //Adapted to screen height, under normal circumstances, the height still uses x.w
    Screentasia().radius(200)    (dart sdk>=2.6 ) another way (200.r)    //Adapt according to the smaller of width or height
    Screentasia().setSp(24)      (dart sdk>=2.6 ) another way (24.sp) //Adapter font
    Screentasia().setPercentage(MediaQuery.sizeOf(context).width,50)   (dart sdk>=2.6 : 50.wp)    //Device width in Percentage, You can use Height just replace MediaQuery.sizeOf(context).width with MediaQuery.sizeOf(context).height
    
    EdgeInsets.all(10).w    //EdgeInsets.all(10.w)
    EdgeInsets.only(left:8,right:8).r // EdgeInsets.only(left:8.r,right:8.r).
    Radius.circular(16).w          //Radius.circular(16.w)
    BorderRadius.all(Radius.circular(16)).w  

    ap(tablet:50) //adaptive percentage, tablet 50 means that adaptive percentage 50% only on tablet screens, the default is 100% for all screens 
```

CHANGELOG.md
# 1.2.0

- update media query

# 1.1.1

- update README


# 1.1.0

- custom screen,features for height


# 1.0.0

- first version


lib/src/init.dart
import 'package:flutter/material.dart';

import '../screentasia.dart';

typedef RebuildFactor = bool Function(MediaQueryData old, MediaQueryData data);

typedef ScreentasiaInitBuilder = Widget Function(
  BuildContext context,
  Widget? child,
);

class RebuildFactors {
  const RebuildFactors._();

  static bool size(MediaQueryData old, MediaQueryData data) {
    return old.size != data.size;
  }
}

class CustomScreen {
  const CustomScreen(this.width, this.height);
  final double width;
  final double height;
}

class AdaptiveFrom {
  const AdaptiveFrom._();

  static CustomScreen customScreen(double width, double height) {
    return CustomScreen(width, height);
  }

  static const CustomScreen mobile = CustomScreen(400, 760);
  static const CustomScreen tablet = CustomScreen(768, 1060);
  static const CustomScreen desktop = CustomScreen(1280, 621);
}

class AdaptivePercentage {
  const AdaptivePercentage(
      {this.mobile = 100, this.tablet = 100, this.desktop = 100});

  final double mobile;
  final double tablet;
  final double desktop;

  double deviceAdaptivePercentage(double screenWidth) {
    if (screenWidth <= 500) {
      return (mobile / 100);
    } else if (screenWidth < 1024 && screenWidth > 500) {
      return (tablet / 100);
    } else {
      return (desktop / 100);
    }
  }
}

class ScreentasiaInit extends StatefulWidget {
  /// A helper widget that initializes [Screentasia]
  const ScreentasiaInit({
    Key? key,
    required this.builder,
    this.adaptiveFrom = AdaptiveFrom.mobile,
    this.adaptivePercentage = const AdaptivePercentage(),
    this.child,
    this.rebuildFactor = RebuildFactors.size,
    this.useInheritedMediaQuery = false,
  }) : super(key: key);

  final ScreentasiaInitBuilder builder;
  final Widget? child;
  final bool useInheritedMediaQuery;
  final RebuildFactor rebuildFactor;
  final CustomScreen adaptiveFrom;
  final AdaptivePercentage adaptivePercentage;

  @override
  State<ScreentasiaInit> createState() => ScreentasiaInitState();
}

class ScreentasiaInitState extends State<ScreentasiaInit>
    with WidgetsBindingObserver {
  MediaQueryData? _mediaQueryData;

  bool wrappedInMediaQuery = false;

  WidgetsBinding get binding => WidgetsFlutterBinding.ensureInitialized();

  MediaQueryData get mediaQueryData => _mediaQueryData!;

  MediaQueryData get newData {
    if (widget.useInheritedMediaQuery) {
      final data = MediaQuery.maybeOf(context);

      if (data != null) {
        wrappedInMediaQuery = true;
        return data;
      }
    }

    return MediaQueryData.fromView(
        WidgetsBinding.instance.platformDispatcher.views.single);
  }

  Widget get child {
    return widget.builder.call(context, widget.child);
  }

  _updateTree(Element el) {
    el.markNeedsBuild();
    el.visitChildren(_updateTree);
  }

  @override
  void initState() {
    super.initState();

    binding.addObserver(this);
  }

  @override
  void didChangeMetrics() {
    final old = _mediaQueryData!;
    final data = newData;

    if (widget.rebuildFactor(old, data)) {
      _mediaQueryData = data;
      _updateTree(context as Element);
    }
  }

  @override
  void didChangeDependencies() {
    super.didChangeDependencies();
    if (_mediaQueryData == null) _mediaQueryData = newData;
    didChangeMetrics();
  }

  @override
  void dispose() {
    binding.removeObserver(this);
    super.dispose();
  }

  @override
  Widget build(BuildContext _context) {
    if (mediaQueryData.size == Size.zero) return const SizedBox.shrink();
    if (!wrappedInMediaQuery) {
      return MediaQuery(
        data: mediaQueryData,
        child: Builder(
          builder: (__context) {
            final deviceSize = MediaQuery.sizeOf(context);

            Screentasia.init(
                __context, widget.adaptiveFrom, widget.adaptivePercentage);
            return AppWidget(deviceSize: deviceSize, child: child);
          },
        ),
      );
    }

    Screentasia.init(_context, widget.adaptiveFrom, widget.adaptivePercentage);

    final deviceSize = MediaQuery.sizeOf(context);
    return AppWidget(deviceSize: deviceSize, child: child);
  }
}


lib/src/extensions.dart
import '../screentasia.dart';
import 'package:flutter/material.dart';

extension SizeExtension on num {
  ///Adaptive Width
  double get w => Screentasia().setWidth(this);

  ///Adaptive Height
  double get h => Screentasia().setHeight(this);

  ///Adaptive Radius
  double get r => Screentasia().radius(this);

  ///Adaptive Size (font,icon)
  double get sp => Screentasia().setSp(this);

  ///Width percentage :ex 50.wp is 50% of screen width
  double get wp {
    return Screentasia().setPercentage(Screentasia().screenWidth, this);
  }

  ///Height percentage :ex 50.hp is 50% of screen Height
  double get hp {
    return Screentasia().setPercentage(Screentasia().screenHeight, this);
  }

  ///Adaptive Percentage :ex al(50) is 50% from new adaptive value
  double ap({adaptivePercentage = const AdaptivePercentage()}) {
    return Screentasia().setAdaptivePercentage(this, adaptivePercentage);
  }
}

extension EdgeInsetsExtension on EdgeInsets {
  /// Creates adapt insets using r [SizeExtension].
  EdgeInsets get r => copyWith(
        top: top.r,
        bottom: bottom.r,
        right: right.r,
        left: left.r,
      );

  EdgeInsets get w => copyWith(
        top: top.w,
        bottom: bottom.w,
        right: right.w,
        left: left.w,
      );

  EdgeInsets get h => copyWith(
        top: top.h,
        bottom: bottom.h,
        right: right.h,
        left: left.h,
      );
}

extension BorderRaduisExtension on BorderRadius {
  /// Creates adapt BorderRadius using r [SizeExtension].
  BorderRadius get r => copyWith(
        bottomLeft: bottomLeft.r,
        bottomRight: bottomRight.r,
        topLeft: topLeft.r,
        topRight: topRight.r,
      );

  BorderRadius get w => copyWith(
        bottomLeft: bottomLeft.w,
        bottomRight: bottomRight.w,
        topLeft: topLeft.w,
        topRight: topRight.w,
      );

  BorderRadius get h => copyWith(
        bottomLeft: bottomLeft.h,
        bottomRight: bottomRight.h,
        topLeft: topLeft.h,
        topRight: topRight.h,
      );
}

extension RaduisExtension on Radius {
  /// Creates adapt Radius using r [SizeExtension].
  Radius get r => Radius.elliptical(x.r, y.r);

  Radius get w => Radius.elliptical(x.w, y.w);

  Radius get h => Radius.elliptical(x.h, y.h);
}


lib/src/screentasia.dart
import '../screentasia.dart';
import 'package:flutter/material.dart';

class Screentasia {
  static final Screentasia _instance = Screentasia._();

  Screentasia._();

  factory Screentasia() {
    return _instance;
  }

  BuildContext? _context;
  late double _screenWidth;
  late double _screenHeight;
  late double _adaptiveFromWidth;
  late double _adaptiveFromHeight;

  late double _deviceAdaptivePercentage;

  double? _lastScreenHeight;

  double get screenWidth => _screenWidth;

  double get screenHeight => _screenHeight;

  double get lastScreenHeight =>
      _lastScreenHeight == null ? screenHeight : _lastScreenHeight!;

  double fullAdaptivePercentageValue(num value) {
    return value * 1 / _deviceAdaptivePercentage;
  }

  double get adaptiveFromHeight1PixelScale => (1 * 100) / _adaptiveFromHeight;

  double get adaptiveFromWidth1PixelScale => (1 * 100) / _adaptiveFromWidth;

  double get currentScreen1PixelScaleWidth =>
      (adaptiveFromWidth1PixelScale * screenWidth) / 100;

  double get currentScreen1PixelScaleHeight {
    if (lastScreenHeight != screenHeight) {
      return (adaptiveFromHeight1PixelScale * screenHeight) / 100;
    }

    _lastScreenHeight = screenHeight;

    return 1;
  }

  double get finalWidthAdaptiveValue => _deviceAdaptivePercentage != 0
      ? currentScreen1PixelScaleWidth * _deviceAdaptivePercentage
      : 1;

  double get finalHeightAdaptiveValue => _deviceAdaptivePercentage != 0
      ? currentScreen1PixelScaleHeight * _deviceAdaptivePercentage
      : 1;

  /// Initializing the Screentasia
  static Future<void> init(
    BuildContext context,
    CustomScreen adaptiveFrom,
    AdaptivePercentage adaptivePercentage,
  ) async {
    final navigatorContext = Navigator.maybeOf(context)?.context as Element?;
    final mediaQueryContext =
        navigatorContext?.getElementForInheritedWidgetOfExactType<MediaQuery>();

    final initCompleter = Completer<void>();

    WidgetsFlutterBinding.ensureInitialized().addPostFrameCallback((_) {
      mediaQueryContext?.visitChildElements((el) => _instance._context = el);
      if (_instance._context != null) initCompleter.complete();
    });

    final deviceData = MediaQuery.sizeOf(context).nonEmptySizeOrNull();

    final deviceSize = deviceData;

    _instance
      .._screenWidth = deviceSize!.width
      .._screenHeight = deviceSize.height
      .._context = context
      .._adaptiveFromHeight = adaptiveFrom.height
      .._adaptiveFromWidth = adaptiveFrom.width
      .._deviceAdaptivePercentage =
          adaptivePercentage.deviceAdaptivePercentage(deviceSize.width);

    return initCompleter.future;
  }

  double setWidth(num width) {
    return width * finalWidthAdaptiveValue;
  }

  double setHeight(num height) => height * finalHeightAdaptiveValue;

  double radius(num r) => r * finalWidthAdaptiveValue;

  double setSp(num fontSize) {
    return fontSize * finalWidthAdaptiveValue;
  }

  double setPercentage(num screenValue, num percentage) {
    try {
      return (percentage > 100 || percentage <= 0)
          ? throw (const PercentageException(
              'The value must be (value > 0 && value <= 100'))
          : screenValue * (percentage / 100);
    } on PercentageException catch (r) {
      throw Exception(r.message);
    }
  }

  double setAdaptivePercentage(
      num value, AdaptivePercentage adaptivePercentage) {
    return fullAdaptivePercentageValue(value) *
        (adaptivePercentage.deviceAdaptivePercentage(screenWidth));
  }
}

extension on Size? {
  Size? nonEmptySizeOrNull() {
    if (this?.isEmpty ?? true) {
      return null;
    } else {
      return this;
    }
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:screentasia/screentasia.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return ScreentasiaInit(
      adaptiveFrom: AdaptiveFrom.mobile,
      adaptivePercentage:
          const AdaptivePercentage(mobile: 100, desktop: 100, tablet: 100),
      builder: (context, child) {
        return MaterialApp(
            debugShowCheckedModeBanner: false,
            title: 'Flutter screentasia',
            home: Scaffold(
                appBar: AppBar(
                  title: const Text('Flutter screentasia Demo  Page'),
                ),
                body: Container(
                    color: Colors.blue,
                    width: 100.wp.ap(
                        adaptivePercentage: const AdaptivePercentage(
                            desktop: 100, tablet: 50, mobile: 25)))));
      },
    );
  }
}
